src/imputors/drop.jl

```diff
@@ -1,73 +1,143 @@
+struct DropObs <: Imputor
+    vardim::Int
+    context::AbstractContext
+end
 
-Removes missing values from the `AbstractArray` or `DataFrame` provided.
 """
-struct Drop <: Imputor end
+    DropObs(; vardim=2, context=Context)
 
-"""
-    impute!(imp::Drop, ctx::Context, data::AbstractVector)
+Removes missing observations from the `AbstractArray` or `Tables.table` provided.
 
-Uses `filter!` to remove missing elements from the array.
+# Keyword Arguments
+* `vardim=2::Int`: Specify the dimension for variables in matrix input data
+* `context::AbstractContext=Context()`: A context which keeps track of missing data
+  summary information
 
-# Arguments
-* `imp::Drop`: this `Imputor` method
-* `ctx::Context`: contextual information for missing data
-* `data::AbstractVector`: the data to impute
+# Example
+```jldoctest
+julia> using Impute: DropObs, Context, impute
 
-# Returns
-* `AbstractVector`: our data array with missing elements removed
+julia> M = [1.0 2.0 missing missing 5.0; 1.1 2.2 3.3 missing 5.5]
+2×5 Array{Union{Missing, Float64},2}:
+ 1.0  2.0   missing  missing  5.0
+ 1.1  2.2  3.3       missing  5.5
+
+julia> impute(M, DropObs(; vardim=1, context=Context(; limit=1.0)))
+2×3 Array{Union{Missing, Float64},2}:
+ 1.0  2.0  5.0
+ 1.1  2.2  5.5
+```
 """
-function impute!(imp::Drop, ctx::Context, data::AbstractVector)
-    return filter!(x -> !ismissing(ctx, x), data)
+DropObs(; vardim=2, context=Context()) = DropObs(vardim, context)
+
+function impute!(data::AbstractVector, imp::DropObs)
+    imp.context() do c
+        filter!(x -> !ismissing(c, x), data)
+    end
 end
 
-"""
-    impute!(imp::Drop, ctx::Context, data::AbstractMatrix)
+function impute!(data::AbstractMatrix, imp::DropObs)
+    imp.context() do c
+        return filterobs(imp, data) do obs
+            !ismissing(c, obs)
+        end
+    end
+end
 
-Finds the missing rows in the matrix and uses a mask (Vector{Bool}) to return the
-`data` with those rows removed. Unfortunately, the mask approach requires copying the matrix.
+# Deleting elements from subarrays doesn't work so we need to collect that data into
+# a separate array.
+impute!(data::SubArray, imp::DropObs) = impute!(collect(data), imp::DropObs)
 
-NOTES (or premature optimizations):
-* We use `view`, but this will change the type of the `data` by returning a `SubArray`
-* We might be able to do something clever by:
-    1. reshaping the data to a vector
-    2. running `deleteat!` for the appropriate indices and
-    3. reshaping the data back to the desired shape.
+function impute!(table, imp::DropObs)
+    imp.context() do c
+        @assert istable(table)
+        rows = Tables.rows(table)
 
-# Arguments
-* `imp::Drop`: this `Imputor` method
-* `ctx::Context`: contextual information for missing data
-* `data::AbstractMatrix`: the data to impute
+        # Unfortunately, we'll need to construct a new table
+        # since Tables.rows is just an iterator
+        table = Iterators.filter(rows) do r
```

Comment:
```suggestion
        new_table = Iterators.filter(rows) do r
```

Reply:
We're intentionally assigning it back because this method signature indicates that it should be mutating.